[dist_optim] fix the bug of none grads on functional optimizers

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #62249

parameter and grads passed to torch.optim.functional should always match, we should skip the parameters that have none gradients to avoid the size mismatch

Differential Revision: [D29929653](https://our.internmc.facebook.com/intern/diff/D29929653/)

Commit:
[dist_optim] fix the bug of none grads on sgd

parameter and grads passed to torch.optim.functional should always match, we should skip the parameters that have none gradients to avoid the size mismatch

Differential Revision: [D29929653](https://our.internmc.facebook.com/intern/diff/D29929653/)

[ghstack-poisoned]

mrshenli commented on torch/testing/_internal/distributed/rpc/dist_optimizer_test.py:
```diff
@@ -209,3 +209,67 @@ def test_dist_optim(self):
+
+        remote_module1 = rpc.remote(owner1, MyModule)
+        remote_module2 = rpc.remote(owner2, MyModule, args=(False,))
+        remote_param1 = remote_method(MyModule.get_w, remote_module1)
```
nit: this can be simplified to `remote_module1.remote().get_w()`?

mrshenli commented on torch/testing/_internal/distributed/rpc/dist_optimizer_test.py:
```diff
@@ -209,3 +209,67 @@ def test_dist_optim(self):
+        remote_param1 = remote_method(MyModule.get_w, remote_module1)
+        remote_param2 = remote_method(MyModule.get_w, remote_module2)
+
+        old_w1_remote = remote_param1.to_here()
```
`old_w1_remote` not used?

mrshenli commented on torch/testing/_internal/distributed/rpc/dist_optimizer_test.py:
```diff
@@ -209,3 +209,67 @@ def test_dist_optim(self):
+            g_cpu.manual_seed(0)
+            t1 = torch.rand((3, 3), requires_grad=True, generator=g_cpu)
+            t2 = torch.rand((3, 3), requires_grad=True, generator=g_cpu)
+            output1 = rpc_async_method(MyModule.forward, remote_module1, t2)
```
nit: this can be `remote_module1.rpc_async().forward(t2)`?

mrshenli commented on torch/testing/_internal/distributed/rpc/dist_optimizer_test.py:
```diff
@@ -209,3 +209,67 @@ def test_dist_optim(self):
+            dist_autograd.backward(context_id, [loss.sum()])
+            dist_optim.step(context_id)
+
+            new_w1 = rpc_async_method(MyModule.get_w, remote_module1).wait()
```
ditto

mrshenli reviewed (approved):
LGTM! Left some minor comments.

mrshenli reviewed (commented):
BTW, looks like we need the same fix for `functional_rprop.py` as well

https://github.com/pytorch/pytorch/blob/a3be2ecc3aee3b0ea0a27f803b8e6ac3245c6789/torch/distributed/optim/functional_rprop.py#L40-L71

andwgu commented:
Thanks for patching this quickly!

Commit:
Update on "[dist_optim] fix the bug of none grads on sgd"


parameter and grads passed to torch.optim.functional should always match, we should skip the parameters that have none gradients to avoid the size mismatch

Differential Revision: [D29929653](https://our.internmc.facebook.com/intern/diff/D29929653/)

[ghstack-poisoned]

Commit:
Update on "[dist_optim] fix the bug of none grads on functional optimizers"


parameter and grads passed to torch.optim.functional should always match, we should skip the parameters that have none gradients to avoid the size mismatch

Differential Revision: [D29929653](https://our.internmc.facebook.com/intern/diff/D29929653/)

[ghstack-poisoned]
